### PR TITLE
chore: warn when NODE_ENV == "production"

### LINKED ISSUE #575: Print a warning message when `NODE_ENV` is `production`
When `NODE_ENV` is `production`, `devDependencies` are not installed.
This can lead to situations where the build fails due to missing dependencies, which results in bug reports and support tickets.

Conceptually, it seems to me builds should never use `NODE_ENV="production"`. Are there any legitimate use cases for this? It seems like most builds would need development dependencies and are technically part of the development flow, not production.

What are your thoughts on printing a warning message when `NODE_ENV` is `production`, warning that development dependencies will not be installed and might be missing?

cc @kaganjd

### PR DESCRIPTION
**Pull request by aitkenster**
#### Summary

Fixes #575

---

For us to review and ship your PR efficiently, please perform the following steps:

- [x] Open a [bug/issue](https://github.com/netlify/build-image/issues/new/choose) before writing your code 🧑‍💻. This ensures we can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
- [x] Read the [contribution guidelines](../CONTRIBUTING.md) 📖. This ensures your code follows our style guide and passes our tests.
- [x] Make sure the status checks below are successful ✅

**A picture of a cute animal (not mandatory, but encouraged)**

image

### CODE REVIEW

File: run-build-functions.sh [hunk 1]
```diff
@@ -281,6 +281,11 @@ install_dependencies() {
     fi
   fi
 
+  if [ "$NODE_ENV" == "production" ]
+  then
+     warn "NODE_ENV is set to 'production'. Any devDependencies in package.json will not be installed"
```
Comment (ehmicky) [inline run-build-functions.sh]: What are your thoughts on the following?

```suggestion
     warn "The environment variable 'NODE_ENV' is set to 'production'. Any 'devDependencies' in package.json will not be installed."
```

File: run-build-functions.sh [hunk 2]
```diff
@@ -888,3 +893,7 @@ unset_go_import_path() {
     unlink $GOPATH/src/$GO_IMPORT_PATH
   fi
 }
+
+warn() {
+  echo "WARNING: $1"
+}
```
Comment (ehmicky) [inline run-build-functions.sh]: [sand] Some IDEs automatically insert a final newline to each file to our [`.editorconfig` file](https://github.com/netlify/build-image/blob/bac1417dd652181d67060aaf756418e6d8ca485a/.editorconfig#L11).
I.e. an unrelated PR might add a newline there.

File: run-build-functions.sh [hunk 3]
```diff
@@ -281,6 +281,11 @@ install_dependencies() {
     fi
   fi
 
+  if [ "$NODE_ENV" == "production" ]
```
Comment (ehmicky) [inline run-build-functions.sh]: Should we put this right before `run_npm`/`run_yarn` instead? Like this, we will not print this when the site has no `package.json`. Granted though that having a `NODE_ENV` environment variable but no `package.json` should be unlikely.

**Comment by ehmicky** [comment]
Thanks @aitkenster!

This looks good to me, except for some small details.

After shipping, we should monitor on Humio how many sites are printing this warning, just in case there happened to be a legitimate reason why some users might need to use `NODE_ENV=production` (which is unclear at the moment).

File: run-build-functions.sh (hunk 1 above)
Comment (aitkenster) [reply run-build-functions.sh]: Yes, I'll change it 👍 I think it's better without the final "." though to be consistent with the rest of the file.

File: run-build-functions.sh (hunk 2 above)
Comment (ehmicky) [reply run-build-functions.sh]: Note: link to setups for `.editorconfig` for different IDEs: https://editorconfig.org/#pre-installed

File: run-build-functions.sh (hunk 1 above)
Comment (ehmicky) [reply run-build-functions.sh]: Good point! :+1: (pun intended)

**Review by ehmicky — APPROVED** [review]

**Pull request merged**